File: go/server/coding_tools_test.go
package server

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"

	core "github.com/SmooAI/smooth-operator-core/go/core"
)

// toolByName pulls a tool out of the CodingTools set by name (they're addressed by name at
// dispatch, so the test does too).
func toolByName(t *testing.T, tools []core.Tool, name string) core.Tool {
	t.Helper()
	for _, tl := range tools {
		if tl.Name() == name {
			return tl
		}
	}
	t.Fatalf("tool %q not found", name)
	return nil
}

func TestCodingToolsSet(t *testing.T) {
	tools := CodingTools(t.TempDir())
	want := []string{"read_file", "write_file", "edit_file", "list_files", "grep", "bash"}
	if len(tools) != len(want) {
		t.Fatalf("got %d tools, want %d", len(tools), len(want))
	}
	for _, n := range want {
		toolByName(t, tools, n) // fatals if missing
	}
}

func TestWriteThenReadFile(t *testing.T) {
	ws := t.TempDir()
	tools := CodingTools(ws)
	ctx := context.Background()

	res, err := toolByName(t, tools, "write_file").Execute(ctx, map[string]any{"path": "hello.txt", "content": "WORLD"})
	if err != nil {
		t.Fatalf("write_file: %v", err)
	}
	if !strings.Contains(res, "Wrote") {
		t.Errorf("write result = %q, want a write confirmation", res)
	}
	// The file actually landed on disk with the exact bytes.
	got, err := os.ReadFile(filepath.Join(ws, "hello.txt"))
	if err != nil {
		t.Fatalf("read back: %v", err)
	}
	if string(got) != "WORLD" {
		t.Errorf("file content = %q, want %q", got, "WORLD")
	}
	// read_file returns line-numbered content.
	rd, err := toolByName(t, tools, "read_file").Execute(ctx, map[string]any{"path": "hello.txt"})
	if err != nil {
		t.Fatalf("read_file: %v", err)
	}
	if !strings.Contains(rd, "WORLD") {
		t.Errorf("read_file = %q, want it to contain WORLD", rd)
	}
}

func TestWriteFileCreatesParentDirs(t *testing.T) {
	ws := t.TempDir()
	tools := CodingTools(ws)
	if _, err := toolByName(t, tools, "write_file").Execute(context.Background(),
		map[string]any{"path": "a/b/c.txt", "content": "deep"}); err != nil {
		t.Fatalf("write nested: %v", err)
	}
	if got, _ := os.ReadFile(filepath.Join(ws, "a/b/c.txt")); string(got) != "deep" {
		t.Errorf("nested file content = %q, want %q", got, "deep")
	}
}

func TestEditFile(t *testing.T) {
	ws := t.TempDir()
	tools := CodingTools(ws)
	ctx := context.Background()
	os.WriteFile(filepath.Join(ws, "f.txt"), []byte("foo bar foo"), 0o644)

	// Non-unique old_string without replace_all is refused.
	if _, err := toolByName(t, tools, "edit_file").Execute(ctx,
		map[string]any{"path": "f.txt", "old_string": "foo", "new_string": "X"}); err == nil {
		t.Error("expected error editing a non-unique string without replace_all")
	}
	// replace_all rewrites every occurrence.
	if _, err := toolByName(t, tools, "edit_file").Execute(ctx,
		map[string]any{"path": "f.txt", "old_string": "foo", "new_string": "X", "replace_all": true}); err != nil {
		t.Fatalf("edit replace_all: %v", err)
	}
	if got, _ := os.ReadFile(filepath.Join(ws, "f.txt")); string(got) != "X bar X" {
		t.Errorf("after edit = %q, want %q", got, "X bar X")
	}
	// Missing old_string is an error.
	if _, err := toolByName(t, tools, "edit_file").Execute(ctx,
		map[string]any{"path": "f.txt", "old_string": "nope", "new_string": "Y"}); err == nil {
		t.Error("expected error when old_string absent")
	}
}

func TestListAndGrep(t *testing.T) {
	ws := t.TempDir()
	tools := CodingTools(ws)
	ctx := context.Background()
	os.WriteFile(filepath.Join(ws, "one.txt"), []byte("needle here\nplain"), 0o644)
	os.MkdirAll(filepath.Join(ws, "node_modules", "junk"), 0o755)
	os.WriteFile(filepath.Join(ws, "node_modules", "junk", "x.txt"), []byte("needle"), 0o644)

	ls, err := toolByName(t, tools, "list_files").Execute(ctx, map[string]any{})
	if err != nil {
		t.Fatalf("list_files: %v", err)
	}
	if !strings.Contains(ls, "one.txt") {
		t.Errorf("list = %q, want one.txt", ls)
	}
	if strings.Contains(ls, "node_modules") {
		t.Errorf("list = %q, should skip node_modules", ls)
	}

	gr, err := toolByName(t, tools, "grep").Execute(ctx, map[string]any{"pattern": "needle"})
	if err != nil {
		t.Fatalf("grep: %v", err)
	}
	if !strings.Contains(gr, "one.txt:1:") {
		t.Errorf("grep = %q, want a one.txt:1 match", gr)
	}
	if strings.Contains(gr, "node_modules") {
		t.Errorf("grep = %q, should skip node_modules", gr)
	}
}

func TestBashRunsInWorkspace(t *testing.T) {
	ws := t.TempDir()
	tools := CodingTools(ws)
	res, err := toolByName(t, tools, "bash").Execute(context.Background(),
		map[string]any{"command": "echo hi > out.txt && cat out.txt"})
	if err != nil {
		t.Fatalf("bash: %v", err)
	}
	if !strings.Contains(res, "exit: 0") || !strings.Contains(res, "hi") {
		t.Errorf("bash result = %q, want exit 0 + hi", res)
	}
	if got, _ := os.ReadFile(filepath.Join(ws, "out.txt")); strings.TrimSpace(string(got)) != "hi" {
		t.Errorf("bash cwd wrong: out.txt = %q", got)
	}
}

func TestBashBlocksCatastrophic(t *testing.T) {
	tools := CodingTools(t.TempDir())
	res, err := toolByName(t, tools, "bash").Execute(context.Background(),
		map[string]any{"command": "rm -rf /"})
	if err != nil {
		t.Fatalf("bash: %v", err)
	}
	if !strings.Contains(res, "BLOCKED") {
		t.Errorf("bash result = %q, want BLOCKED for rm -rf /", res)
	}
}

func TestPathConfinement(t *testing.T) {
	ws := t.TempDir()
	// Absolute escape.
	if _, err := resolveWorkspacePath(ws, "/etc/passwd"); err == nil {
		t.Error("expected /etc/passwd to be rejected")
	}
	// Relative escape.
	if _, err := resolveWorkspacePath(ws, "../../etc/passwd"); err == nil {
		t.Error("expected ../../etc/passwd to be rejected")
	}
	// Empty.
	if _, err := resolveWorkspacePath(ws, ""); err == nil {
		t.Error("expected empty path to be rejected")
	}
	// Contained relative + absolute-within both allowed.
	if _, err := resolveWorkspacePath(ws, "sub/file.txt"); err != nil {
		t.Errorf("contained relative rejected: %v", err)
	}
	if _, err := resolveWorkspacePath(ws, filepath.Join(ws, "in.txt")); err != nil {
		t.Errorf("absolute-within-workspace rejected: %v", err)
	}
}

func TestWriteFileRejectsEscape(t *testing.T) {
	ws := t.TempDir()
	tools := CodingTools(ws)
	if _, err := toolByName(t, tools, "write_file").Execute(context.Background(),
		map[string]any{"path": "../escape.txt", "content": "x"}); err == nil {
		t.Error("write_file should reject a path escaping the workspace")
	}
}

File: typescript/server/test/codingTools.test.ts
/**
 * Coding-toolset tests (th-82ad57) — the TS mirror of `go/server/coding_tools_test.go`.
 * Covers the six-tool set, each tool's happy + error path, and (the security-critical bit)
 * that the workspace path guard rejects escapes.
 */
import { mkdtemp, mkdir, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import * as path from 'node:path';
import { afterEach, describe, expect, it } from 'vitest';

import { codingTools, resolveWorkspacePath } from '../src/codingTools.js';

const created: string[] = [];

async function tempWorkspace(): Promise<string> {
    // realpath-ish: macOS /var → /private/var, and the guard compares resolved paths, so the
    // workspace root must be the same string the tools resolve to.
    const dir = await mkdtemp(path.join(await realTmp(), 'coding-tools-'));
    created.push(dir);
    return dir;
}

async function realTmp(): Promise<string> {
    const { realpath } = await import('node:fs/promises');
    return realpath(tmpdir());
}

afterEach(async () => {
    await Promise.all(created.splice(0).map((d) => rm(d, { recursive: true, force: true })));
});

function toolByName(workspace: string, name: string) {
    const tool = codingTools(workspace).find((t) => t.name === name);
    if (!tool) throw new Error(`tool "${name}" not found`);
    return tool;
}

describe('codingTools', () => {
    it('exposes exactly the six parity tools', async () => {
        const tools = codingTools(await tempWorkspace());
        expect(tools.map((t) => t.name).sort()).toEqual(['bash', 'edit_file', 'grep', 'list_files', 'read_file', 'write_file']);
    });

    it('writes then reads a file', async () => {
        const ws = await tempWorkspace();
        const res = await toolByName(ws, 'write_file').execute({ path: 'hello.txt', content: 'WORLD' });
        expect(res).toContain('Wrote');
        expect(await readFile(path.join(ws, 'hello.txt'), 'utf8')).toBe('WORLD');

        const rd = await toolByName(ws, 'read_file').execute({ path: 'hello.txt' });
        expect(rd).toContain('WORLD');
        expect(rd).toContain('     1\t'); // line-numbered
    });

    it('honours the read_file line window', async () => {
        const ws = await tempWorkspace();
        await writeFile(path.join(ws, 'many.txt'), 'a\nb\nc\nd\n');
        const rd = await toolByName(ws, 'read_file').execute({ path: 'many.txt', offset: 2, limit: 2 });
        expect(rd).toContain('b');
        expect(rd).toContain('c');
        expect(rd).not.toContain('\ta\n');
        expect(rd).not.toContain('d');
    });

    it('read_file errors on a missing file', async () => {
        const ws = await tempWorkspace();
        await expect(toolByName(ws, 'read_file').execute({ path: 'nope.txt' })).rejects.toThrow(/read nope.txt/);
    });

    it('write_file creates parent dirs', async () => {
        const ws = await tempWorkspace();
        await toolByName(ws, 'write_file').execute({ path: 'a/b/c.txt', content: 'deep' });
        expect(await readFile(path.join(ws, 'a/b/c.txt'), 'utf8')).toBe('deep');
    });

    it('edits a file, refusing non-unique matches without replace_all', async () => {
        const ws = await tempWorkspace();
        await writeFile(path.join(ws, 'f.txt'), 'foo bar foo');
        const edit = toolByName(ws, 'edit_file');

        await expect(edit.execute({ path: 'f.txt', old_string: 'foo', new_string: 'X' })).rejects.toThrow(/occurs 2 times/);

        await edit.execute({ path: 'f.txt', old_string: 'foo', new_string: 'X', replace_all: true });
        expect(await readFile(path.join(ws, 'f.txt'), 'utf8')).toBe('X bar X');

        await expect(edit.execute({ path: 'f.txt', old_string: 'nope', new_string: 'Y' })).rejects.toThrow(/not found/);
    });

    it('lists and greps, skipping node_modules', async () => {
        const ws = await tempWorkspace();
        await writeFile(path.join(ws, 'one.txt'), 'needle here\nplain');
        await mkdir(path.join(ws, 'node_modules', 'junk'), { recursive: true });
        await writeFile(path.join(ws, 'node_modules', 'junk', 'x.txt'), 'needle');

        const ls = await toolByName(ws, 'list_files').execute({});
        expect(ls).toContain('one.txt');
        expect(ls).not.toContain('node_modules');

        const gr = await toolByName(ws, 'grep').execute({ pattern: 'needle' });
        expect(gr).toContain('one.txt:1:');
        expect(gr).not.toContain('node_modules');
    });

    it('grep reports no matches and rejects an invalid regexp', async () => {
        const ws = await tempWorkspace();
        await writeFile(path.join(ws, 'one.txt'), 'plain');
        expect(await toolByName(ws, 'grep').execute({ pattern: 'zzz' })).toContain('(no matches');
        await expect(toolByName(ws, 'grep').execute({ pattern: '([' })).rejects.toThrow(/invalid regexp/);
    });

    it('runs bash in the workspace', async () => {
        const ws = await tempWorkspace();
        const res = await toolByName(ws, 'bash').execute({ command: 'echo hi > out.txt && cat out.txt' });
        expect(res).toContain('exit: 0');
        expect(res).toContain('hi');
        expect((await readFile(path.join(ws, 'out.txt'), 'utf8')).trim()).toBe('hi');
    });

    it('reports a non-zero bash exit code', async () => {
        const ws = await tempWorkspace();
        expect(await toolByName(ws, 'bash').execute({ command: 'exit 3' })).toContain('exit: 3');
    });

    it('kills a bash command that exceeds its timeout', async () => {
        const ws = await tempWorkspace();
        const res = await toolByName(ws, 'bash').execute({ command: 'sleep 5', timeout: 1 });
        expect(res).toContain('killed (timeout after 1s)');
    });

    it('blocks catastrophic bash commands', async () => {
        const ws = await tempWorkspace();
        expect(await toolByName(ws, 'bash').execute({ command: 'rm -rf /' })).toContain('BLOCKED');
    });

    it('requires the documented arguments', async () => {
        const ws = await tempWorkspace();
        await expect(toolByName(ws, 'read_file').execute({})).rejects.toThrow(/missing required argument "path"/);
        await expect(toolByName(ws, 'write_file').execute({ path: 'x', content: 5 })).rejects.toThrow(/must be a string/);
    });
});

describe('resolveWorkspacePath (trust boundary)', () => {
    it('rejects escapes and accepts contained paths', async () => {
        const ws = await tempWorkspace();
        expect(() => resolveWorkspacePath(ws, '/etc/passwd')).toThrow(/outside the workspace/);
        expect(() => resolveWorkspacePath(ws, '../../etc/passwd')).toThrow(/outside the workspace/);
        expect(() => resolveWorkspacePath(ws, 'sub/../../escape')).toThrow(/outside the workspace/);
        expect(() => resolveWorkspacePath(ws, `${ws}/../sibling`)).toThrow(/outside the workspace/);
        expect(() => resolveWorkspacePath(ws, '')).toThrow(/empty path/);

        expect(resolveWorkspacePath(ws, 'sub/file.txt')).toBe(path.join(ws, 'sub/file.txt'));
        expect(resolveWorkspacePath(ws, path.join(ws, 'in.txt'))).toBe(path.join(ws, 'in.txt'));
        expect(resolveWorkspacePath(ws, './nested/../ok.txt')).toBe(path.join(ws, 'ok.txt'));
    });

    it('is enforced by the tools themselves, not just the helper', async () => {
        const ws = await tempWorkspace();
        await expect(toolByName(ws, 'write_file').execute({ path: '../escape.txt', content: 'x' })).rejects.toThrow(/outside the workspace/);
        await expect(toolByName(ws, 'read_file').execute({ path: '/etc/passwd' })).rejects.toThrow(/outside the workspace/);
        await expect(toolByName(ws, 'edit_file').execute({ path: '../f.txt', old_string: 'a', new_string: 'b' })).rejects.toThrow(
            /outside the workspace/,
        );
        await expect(toolByName(ws, 'list_files').execute({ path: '..' })).rejects.toThrow(/outside the workspace/);
        await expect(toolByName(ws, 'grep').execute({ pattern: 'x', path: '/etc' })).rejects.toThrow(/outside the workspace/);
    });
});
